src/index/utxoscriptindex.cpp

```diff
@@ -0,0 +1,455 @@
+    unsigned int GetValueSize() const;
+
+    bool Valid() const;
+    void Next();
+
+    std::pair<char, ScriptHash> keyTmp;
+    std::unique_ptr<CDBIterator> pcursor;
+};
+
+class UtxoScriptIndex::DB : public BaseIndex::DB
+{
+public:
+    explicit DB(size_t nCacheSize, bool fMemory = false, bool fWipe = false);
+
+    bool WriteUtxos(std::map<ScriptHash, SerializableUtxoSet>& scriptUtxos);
+    bool ReadUtxos(const ScriptHash&, SerializableUtxoSet& utxoSet);
+
+    UtxoScriptIndexDBCursor* Cursor() const;
+};
+
+UtxoScriptIndex::DB::DB(size_t nCacheSize, bool fMemory, bool fWipe) :
+    BaseIndex::DB(GetDataDir() / "indexes" / "coinsbyscript", nCacheSize, fMemory, fWipe)
+{
+
+}
+
+bool UtxoScriptIndex::DB::WriteUtxos(std::map<ScriptHash, SerializableUtxoSet>& scriptUtxos)
+{
+    CDBBatch batch(*this);
+    for(const auto& scriptUtxo: scriptUtxos)
+    {
+        auto key = std::make_pair(DB_UTXO, std::ref(scriptUtxo.first));
+        auto& value = scriptUtxo.second;
+        if(scriptUtxo.second.empty())
+            batch.Erase(key);
+        else
+            batch.Write(key,value);
+    }
+    return WriteBatch(batch);
+}
+
+bool UtxoScriptIndex::DB::ReadUtxos(const ScriptHash& scriptId, SerializableUtxoSet& utxoSet)
+{
+    return Read(std::make_pair(DB_UTXO, scriptId), utxoSet);
+}
+
+UtxoScriptIndexDBCursor* UtxoScriptIndex::DB::Cursor() const
+{
+    UtxoScriptIndexDBCursor* cursor = new UtxoScriptIndexDBCursor(
+                                    const_cast<CDBWrapper*>(
+                                        static_cast<const CDBWrapper*>(this))->NewIterator());
+    cursor->pcursor->Seek(DB_UTXO);
+    if(!(cursor->pcursor->Valid()))
+        cursor->keyTmp.first = 0;
+    else
+        cursor->pcursor->GetKey(cursor->keyTmp);
+    return cursor;
+}
+
+bool UtxoScriptIndexDBCursor::GetKey(ScriptHash& key)
+{
+    if(keyTmp.first == DB_UTXO)
+    {
+        key = keyTmp.second;
+        return true;
+    }
+    return false;
+}
+
+bool UtxoScriptIndexDBCursor::GetValue(SerializableUtxoSet& utxoSet)
+{
+    return pcursor->GetValue(utxoSet);
+}
+
+bool UtxoScriptIndexDBCursor::Valid() const
+{
+    return keyTmp.first == DB_UTXO;
+}
+
+void UtxoScriptIndexDBCursor::Next()
+{
+    pcursor->Next();
+    if(pcursor->Valid() && pcursor->GetKey(keyTmp))
+        return;
+    else
+        keyTmp.first = 0;
+}
+
+UtxoScriptIndex::UtxoScriptIndex(size_t n_cache_size, bool f_memory, bool f_wipe) :
+    m_db(std::move(MakeUnique<UtxoScriptIndex::DB>(n_cache_size, f_memory, f_wipe)))
+{
+
+}
+
+UtxoScriptIndex::~UtxoScriptIndex() {
+    UnregisterValidationInterface(this);
+}
+
+void UtxoScriptIndex::Start()
+{
+    RegisterValidationInterface(this);
+}
+
+void UtxoScriptIndex::TransactionAddedToMempool(const CTransactionRef &ptxn) {
+    LOCK(cs_utxoCacheMempool);
+    const CTransaction &tx = *ptxn;
+    for (unsigned int i = 0; i < tx.vout.size(); i++)
+    {
+        if (tx.vout[i].IsNull() || tx.vout[i].scriptPubKey.IsUnspendable())
```

Comment:
Is this check needed? In what case will a transaction be added to the mempool where this is true? I might be missing something.